### PR TITLE
Random number generation, two-frame diff steps, adaptive and optimal constant baselines

### PR DESCRIPTION
**Pull request by RenatoGeh**
This patchset does the following:

- Replaces numpy random number generation with torch's.
- Replaces single-frame steps (i.e. Pong's image feed matrices) with a two-frame diff step (i.e. a difference matrix of two consecutive frames).
- Adds an adaptive baseline based on [1].
- Adds the optimal constant baseline with a two-sample approximation of the expected value [2].

**References**

[1] - R. J. Williams. Simple statistical gradient-following algorithms for connectionist reinforcement learning
[2] - J. Peters, S. Schaal. Policy gradient methods for robotics

### CODE REVIEW

File: Implementations/Algorithms/Pong_REINFORCE.py [hunk 1]
```diff
@@ -28,39 +40,37 @@ def set_device(self, device):
...
         I = I[::2,::2,0] # downsample by factor of 2
         I[I == 144] = 0 # erase background (background type 1)
         I[I == 109] = 0 # erase background (background type 2)
         I[I != 0] = 1 # everything else (paddles, ball) just set to 1
         return torch.from_numpy(I.astype(np.float).ravel()).float().unsqueeze(0)
 
     def forward(self, state):
-        state = self.prepro(state).to(self.device)
-        x = F.relu(self.linear1(state))
+        # state = self.prepro(state).to(self.device)
```
Comment (renato-scaroni) [inline Implementations/Algorithms/Pong_REINFORCE.py]: this line should be removed, right?

File: Implementations/Algorithms/Pong_REINFORCE.py [hunk 2]
```diff
@@ -28,39 +40,37 @@ def set_device(self, device):
...
+        x = F.relu(self.linear1(v))
         # TODO: when refactor get_action remember to
         # take this softmax function away
         x = F.softmax(self.linear2(x), dim=1)
         return x
 
 class Agent:
 
-    def __init__(self, env, learning_rate=3e-4, save_plot=True, show_plot=False):
+    def __init__(self, env, learning_rate=1e-3, save_plot=True, show_plot=False, trand=True,
```
Comment (renato-scaroni) [inline Implementations/Algorithms/Pong_REINFORCE.py]: What does this trand parameter do?

File: Implementations/Algorithms/Pong_REINFORCE.py [hunk 3]
```diff
@@ -77,45 +87,93 @@ def __init__(self, env, learning_rate=3e-4, save_plot=True, show_plot=False):
...
+    #  - last log_probs
+    def update_policy(self, rewards, log_probs, last_g, last_b, last_log_probs):
         discounted_rewards = []
+        L = []
+        g0 = 0
 
         for t in range(len(rewards)):
+            gamma = GAMMA
             Gt = 0
             for pw, r in enumerate(rewards[t:]):
```
Comment (renato-scaroni) [inline Implementations/Algorithms/Pong_REINFORCE.py]: I wonder if it would be nicer to remove the enumerate and do for r in rewards[t:]. I think it would be clearer to read, and the way gamma is being computed now makes pw useless

File: Implementations/Algorithms/Pong_REINFORCE.py [hunk 4]
```diff
@@ -28,39 +40,37 @@ def set_device(self, device):
...
+                 baseline=None):
         self.env = env
         self.show_plot = show_plot
         self.save_plot = save_plot
 
         self.num_actions = self.env.action_space.n
-        self.policy_network = PolicyNetwork(6400, self.env.action_space.n)
+        self.policy_network = PolicyNetwork(WIDTH*HEIGHT, self.env.action_space.n)
+
+        self.trand = trand
```
Comment (renato-scaroni) [inline Implementations/Algorithms/Pong_REINFORCE.py]: what this option trand means?

File: Implementations/Algorithms/Pong_REINFORCE.py [hunk 5]
```diff
@@ -159,9 +217,49 @@ def set_seeds(self, s):
         torch.manual_seed(s)
         random.seed(s)
 
+class DiffFrame(gym.ObservationWrapper):
+    def __init__(self, env):
+        super().__init__(env)
+
+    @staticmethod
+    def prepro(I):
```
Comment (renato-scaroni) [inline Implementations/Algorithms/Pong_REINFORCE.py:217]: is this method the same as above? why are they repeated? Shouldn't it be better to organize them on a helper module and keep only one implementation?

File: Implementations/Algorithms/Pong_REINFORCE.py [hunk 6]
```diff
@@ -159,9 +217,49 @@ def set_seeds(self, s):
...
+        # env = atari_wrappers.ScaledFloatFrame(env)
+    # if clip_rewards:
+        # env = atari_wrappers.ClipRewardEnv(env)
+    if frame_stack:
+        env = atari_wrappers.FrameStack(env, 2)
+    return DiffFrame(env)
+
+
+WIDTH = 98
+HEIGHT = 80
```
Comment (renato-scaroni) [inline Implementations/Algorithms/Pong_REINFORCE.py]: For organzation sake it would be better to keep the constants together at the begining of the file.

**Review by renato-scaroni — COMMENTED** [review]
Just some minor changes, generally aiming for better readability and code organization

**Review by renato-scaroni — APPROVED** [review]
